lasagne/layers/local.py

```diff
@@ -0,0 +1,198 @@
+        on its filter and on the input channel at the same channel index.
+        In this case, the number of output channels (i.e. number of filters)
+        should be equal to the number of input channels.
+
+    **kwargs
+        Any additional keyword arguments are passed to the `Layer` superclass.
+
+    Attributes
+    ----------
+    W : Theano shared variable or expression
+        Variable or expression representing the filter weights.
+
+    b : Theano shared variable or expression
+        Variable or expression representing the biases.
+
+    Notes
+    -----
+    This implementation computes the output tensor by iterating over the filter
+    weights and multiplying them with shifted versions of the input tensor.
+    This implementation assumes no stride, 'same' padding and no dilation.
+
+    Raises
+    ------
+    ValueError
+        When ``channelwise`` is set to ``True`` and the number of filters
+        differs from the number of input channels, a `ValueError` is raised.
+    """
+    def __init__(self, incoming, num_filters, filter_size, stride=(1, 1),
+                 pad='same', untie_biases=False,
+                 W=init.GlorotUniform(), b=init.Constant(0.),
+                 nonlinearity=nonlinearities.rectify, flip_filters=True,
+                 channelwise=False, **kwargs):
+        self.channelwise = channelwise
+        super(LocallyConnected2DLayer, self).__init__(
+            incoming, num_filters, filter_size, stride=stride, pad=pad,
+            untie_biases=untie_biases, W=W, b=b, nonlinearity=nonlinearity,
+            flip_filters=flip_filters, **kwargs)
+        # require no stride
+        if self.stride != (1, 1):
+            raise NotImplementedError(
+                "LocallyConnected2DLayer requires stride=1 / (1, 1), but got "
+                "%r." % (stride,))
+        # require same convolution
+        if self.pad != 'same':
+            raise NotImplementedError(
+                "LocallyConnected2DLayer requires pad='same', but got %r." %
+                (pad,))
+
+    def get_W_shape(self):
+        num_input_channels = self.input_shape[1]
+        output_shape = self.get_output_shape_for(self.input_shape)
+        if self.channelwise:
+            if self.channelwise and self.num_filters != num_input_channels:
+                raise ValueError("num_filters and the number of input "
+                                 "channels should match when channelwise is "
+                                 "true, but got num_filters=%r and %d input "
+                                 "channels" %
+                                 (self.num_filters, num_input_channels))
+            return (self.num_filters,) + self.filter_size + output_shape[-2:]
+        else:
+            return (self.num_filters, num_input_channels) + \
+                   self.filter_size + output_shape[-2:]
+
+    def convolve(self, input, **kwargs):
+        output_shape = self.get_output_shape_for(self.input_shape)
```

Comment:
`output_shape = self.output_shape` would be enough here -- we're past the constructor, so this has been set already. It was only required in `get_W_shape()`.